src/constraints/mod.rs

```diff
@@ -0,0 +1,178 @@
+) -> Result<Vec<NonNativeFieldVar<F, CF>>, SynthesisError> {
+    let all_nonnative_bits_le = all_nonnative_bits_le.into_iter().collect::<Vec<_>>();
+    if all_nonnative_bits_le.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    let mut max_nonnative_bits = 0usize;
+    for bits in &all_nonnative_bits_le {
+        max_nonnative_bits = max_nonnative_bits.max(bits.len());
+    }
+
+    let mut lookup_table = Vec::<Vec<CF>>::new();
+    let mut cur = F::one();
+    for _ in 0..max_nonnative_bits {
+        let repr = AllocatedNonNativeFieldVar::<F, CF>::get_limbs_representations(
+            &cur,
+            OptimizationType::Constraints,
+        )?;
+        lookup_table.push(repr);
+        cur.double_in_place();
+    }
+
+    let params = get_params(
+        F::size_in_bits(),
+        CF::size_in_bits(),
+        OptimizationType::Constraints,
+    );
+
+    let mut output = Vec::with_capacity(all_nonnative_bits_le.len());
+    for nonnative_bits_le in all_nonnative_bits_le {
+        let mut val = vec![CF::zero(); params.num_limbs];
+        let mut lc = vec![LinearCombination::<CF>::zero(); params.num_limbs];
+
+        for (j, bit) in nonnative_bits_le.iter().enumerate() {
+            if bit.value().unwrap_or_default() {
+                for (k, val) in val.iter_mut().enumerate().take(params.num_limbs) {
+                    *val += &lookup_table[j][k];
+                }
+            }
+
+            #[allow(clippy::needless_range_loop)]
+            for k in 0..params.num_limbs {
+                lc[k] = &lc[k] + bit.lc() * lookup_table[j][k];
+            }
+        }
+
+        let mut limbs = Vec::new();
+        for k in 0..params.num_limbs {
+            let gadget =
+                AllocatedFp::new_witness(ark_relations::ns!(cs, "alloc"), || Ok(val[k])).unwrap();
+            lc[k] = lc[k].clone() - (CF::one(), gadget.variable);
+            cs.enforce_constraint(lc!(), lc!(), lc[k].clone()).unwrap();
+            limbs.push(FpVar::<CF>::from(gadget));
+        }
+
+        output.push(NonNativeFieldVar::<F, CF>::Var(
+            AllocatedNonNativeFieldVar::<F, CF> {
+                cs: cs.clone(),
+                limbs,
+                num_of_additions_over_normal_form: CF::zero(),
+                is_in_the_normal_form: true,
+                target_phantom: Default::default(),
+            },
+        ));
+    }
+
+    Ok(output)
+}
+
+/// The interface for a cryptographic sponge.
+/// A sponge can `absorb` or take in inputs and later `squeeze` or output bytes or field elements.
+/// The outputs are dependent on previous `absorb` and `squeeze` calls.
+pub trait CryptographicSpongeVar<CF: PrimeField, S: CryptographicSponge<CF>>: Clone {
+    /// Initialize a new instance of the sponge.
+    fn new(cs: ConstraintSystemRef<CF>) -> Self;
+
+    /// Returns a ref to the underlying constraint system the sponge is operating in.
+    fn cs(&self) -> ConstraintSystemRef<CF>;
```

Comment:
I think that we do not need this

Reply:
Ok given the consensus that the struct actually stores state, it's fine to have the `cs` method. However, maybe we want to do this via a `R1CSVar` trait bound?

Reply:
cc @ValarDragon

Reply:
ping